lib/shelley/src/Cardano/Wallet/Shelley/Network.hs

```diff
@@ -0,0 +1,515 @@
+    , ClientStIdle (..)
+    , ClientStIntersect (..)
+    , ClientStNext (..)
+    , chainSyncClientPeer
+    )
+import Ouroboros.Network.Protocol.ChainSync.Codec
+    ( codecChainSync )
+import Ouroboros.Network.Protocol.ChainSync.Type
+    ( ChainSync )
+import Ouroboros.Network.Protocol.Handshake.Version
+    ( DictVersion (..), simpleSingletonVersions )
+import Ouroboros.Network.Protocol.LocalTxSubmission.Client
+    ( LocalTxSubmissionClient (..), localTxSubmissionClientPeer )
+import Ouroboros.Network.Protocol.LocalTxSubmission.Codec
+    ( codecLocalTxSubmission )
+import Ouroboros.Network.Protocol.LocalTxSubmission.Type
+    ( LocalTxSubmission )
+
+import Control.Exception
+    ( catch, throwIO )
+import Control.Monad.Class.MonadST
+    ( MonadST )
+import Control.Monad.Class.MonadThrow
+    ( MonadThrow )
+import Control.Monad.Class.MonadTimer
+    ( MonadTimer )
+import Control.Tracer
+    ( Tracer, contramap )
+import Data.ByteString.Lazy
+    ( ByteString )
+import Data.Text
+    ( Text )
+import Data.Void
+    ( Void )
+import Network.Socket
+    ( AddrInfo (..), Family (..), SockAddr (..), SocketType (..) )
+
+import qualified Cardano.Chain.Block as CC
+import qualified Cardano.Chain.Common as CC
+import qualified Cardano.Chain.UTxO as CC
+import qualified Cardano.Crypto as CC
+import qualified Cardano.Wallet.Primitive.Types as W
+import qualified Codec.Serialise as CBOR
+import qualified Data.ByteArray as BA
+import qualified Data.List.NonEmpty as NE
+import qualified Network.Socket as Socket
+import qualified Ouroboros.Network.Block as Abstract
+import qualified Ouroboros.Network.Point as Point
+
+-- ----------------
+-- Development
+
+import Cardano.BM.Trace
+    ( nullTracer )
+import Control.Monad
+    ( (>=>) )
+
+convertTx :: CC.Tx -> W.Tx
+convertTx tx = W.Tx
+    { W.txId = convertHash $ CC.hash tx
+    , W.resolvedInputs = map convertTxIn $ NE.toList $ CC.txInputs tx
+    , W.outputs = map convertTxOut $ NE.toList $ CC.txOutputs tx
+    }
+  where
+    convertHash = W.Hash . BA.convert
+
+    convertTxIn (CC.TxInUtxo txH idx) = (W.TxIn (convertHash txH) idx, coin)
+       where
+          coin = error "1. We can't get the coin of Byron inputs!"
+
+    convertTxOut o = W.TxOut
+        (convertAddr $ CC.txOutAddress o)
+        (convertCoin $ CC.txOutValue o)
+      where
+        convertAddr = error "todo use toBinary"
+        -- Byron addresses should be no problem. I'm not sure how Shelley
+        -- addresses will look.
+        convertCoin = W.Coin . CC.unsafeGetLovelace
+
+convertBlockHeader :: CC.Block -> W.BlockHeader
+convertBlockHeader b = W.BlockHeader
+    { W.slotId =
+        error "2. can't easily get the SlotId"
+            -- The cardano-ledger equivalent is EpochAndSlotCount, i.e epoch
+            -- number and the slot counted from the start of the epoch.
+            --
+            -- But block headers only contain the slot number (SlotNo), counted
+            -- from genesis.
```

Comment:
So this matches more the idea of "flat" slots that we have right ?

Reply:
Yes. I think this is just good. 

Converting `SlotNo` to `UTCTime` is simpler than from `SlotId`. 

Sometimes we need to be aware of epochs. E.g.
1. To know when  [delegation certificates](https://github.com/input-output-hk/cardano-ledger/blob/dbdb643722e431e4d232345a0eafdc7bdeab7b60/cardano-ledger/src/Cardano/Chain/Delegation/Certificate.hs#L90) take effect.
2. Counting blocks per epoch (if we are still going to do that)

But… we could just use the [`BoundaryBlock`](https://github.com/input-output-hk/cardano-ledger/blob/dbdb643722e431e4d232345a0eafdc7bdeab7b60/cardano-ledger/src/Cardano/Chain/Block/Header.hs#L448-L453) for that, which tell us the new epoch when it starts.

Reply:
Meh. I'd rather also adopt this approach of keeping track of `SlotNo` instead of `SlotId`. There are trade-offs obviously, especially in the way we display things at the API-level but it's probably worth the short.